File: threekt-render/src/main/kotlin/info/laht/krender/threekt/objects/Capsule.kt
package info.laht.krender.threekt.objects

import info.laht.threekt.core.Object3DImpl
import info.laht.threekt.geometries.CylinderBufferGeometry
import info.laht.threekt.geometries.SphereBufferGeometry
import info.laht.threekt.materials.Material
import info.laht.threekt.materials.MeshBasicMaterial
import info.laht.threekt.objects.Mesh

internal class Capsule(
    radius: Float,
    height: Float,
    radiusSegments: Int = 16,
    heightSegments: Int = 8,
    material: Material = MeshBasicMaterial()
) : Object3DImpl() {

    init {

        val body = Mesh(CylinderBufferGeometry(radius, radius, height, radiusSegments, heightSegments), material)
        val upper = Mesh(SphereBufferGeometry(radius, radiusSegments, heightSegments), material)
        val lower = Mesh(SphereBufferGeometry(radius, radiusSegments, heightSegments), material)

        upper.position.set(0, height / 2, 0)
        lower.position.set(0, -height / 2, 0)

        add(body, upper, lower)

    }

}

File: web/src/main/resources/js/Capsule.js
Capsule = function (radius, cylinderHeight, mat, radiusSegments, heightSegments) {
    let capsule = new THREE.Object3D();
    radiusSegments = radiusSegments || 16;
    heightSegments = heightSegments || 8;
    let body = new THREE.Mesh(new THREE.CylinderGeometry(radius, radius, cylinderHeight, radiusSegments, heightSegments, true), mat),
        upperSphere = new THREE.Mesh(new THREE.SphereGeometry(radius, radiusSegments, heightSegments, 0, Math.PI * 2, 0, Math.PI / 2), mat),
        lowerSphere = new THREE.Mesh(new THREE.SphereGeometry(radius, radiusSegments, heightSegments, 0, Math.PI * 2, Math.PI / 2, Math.PI / 2), mat);
    upperSphere.position.set(0, cylinderHeight / 2, 0);
    lowerSphere.position.set(0, -cylinderHeight / 2, 0);
    capsule.add(body);
    capsule.add(upperSphere);
    capsule.add(lowerSphere);
    return capsule;
};
